File: app/src/main/java/com/absinthe/kage/connect/tcp/KageSocket.java
package com.absinthe.kage.connect.tcp;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.Log;

import com.absinthe.kage.connect.protocol.IProtocolHandler;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ConnectException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketTimeoutException;

public class KageSocket {
    private static final String TAG = KageSocket.class.getSimpleName();
    private Socket mSocket;
    private ISocketCallback mSocketCallback;
    private DataInputStream mIn;
    private DataOutputStream mOut;
    private IPacketWriter mPacketWriter;
    private IPacketReader mPacketReader;

    public void connect(final String ip, final int port, int timeout) {
        synchronized (KageSocket.class) {
            if (mSocket == null) {
                new ConnectThread(ip, port, timeout).start();
            } else {
                Log.e(TAG, "Socket is already exist");
            }
        }
    }

    private class ConnectThread extends Thread {

        private String ip;
        private int port;
        private int timeout;

        public ConnectThread(String ip, int port, int timeout) {
            this.ip = ip;
            this.port = port;
            this.timeout = timeout;
        }

        @Override
        public void run() {
            synchronized (KageSocket.class) {
                if (mSocket == null) {
                    try {
                        mSocket = new Socket();
                        SocketAddress endPoint = new InetSocketAddress(ip, port);
                        mSocket.connect(endPoint, timeout);
                        mSocket.setKeepAlive(true);
                        mIn = new DataInputStream(mSocket.getInputStream());
                        mOut = new DataOutputStream(mSocket.getOutputStream());
                        mPacketWriter = new PacketWriter(mOut, mSocketCallback);
                        mPacketReader = new PacketReader(mIn, mSocketCallback);
                        ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                            if (mSocketCallback != null) {
                                mSocketCallback.onConnected();
                            }
                        });
                    } catch (final Exception e) {
                        Log.i(TAG, "Socket connection Exception: " + e);
                        mSocket = null;
                        if (e instanceof SocketTimeoutException) {
                            ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                                if (mSocketCallback != null) {
                                    mSocketCallback.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_TIMEOUT, e);
                                }
                            });
                        } else if (e instanceof ConnectException) {
                            ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                                if (mSocketCallback != null) {
                                    mSocketCallback.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE, e);
                                }
                            });
                        } else {
                            ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                                if (mSocketCallback != null) {
                                    mSocketCallback.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_UNKNOWN, e);
                                }
                            });
                        }
                    }
                }
            }
        }
    }

    public boolean disConnect() {
        synchronized (KageSocket.class) {
            if (mSocket != null && mSocket.isConnected()) {
                try {
                    mIn.close();
                    mOut.close();
                    mSocket.close();
                    mIn = null;
                    mOut = null;
                    mSocket = null;
                    if (null != mPacketWriter) {
                        mPacketWriter.shutdown();
                        mPacketWriter = null;
                    }
                    if (null != mPacketReader) {
                        mPacketReader.shutdown();
                        mPacketReader = null;
                    }
                    ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                        if (mSocketCallback != null) {
                            mSocketCallback.onDisConnected();
                        }
                    });
                    return true;
                } catch (IOException e) {
                    Log.e(TAG, e.toString());
                    return false;
                }
            }
            return false;
        }
    }

    public void send(final Packet packet) {
        synchronized (KageSocket.class) {
            if (mPacketWriter != null) {
                if (packet instanceof Request) {
                    Request request = (Request) packet;
                    request.setId(System.currentTimeMillis() + "");
                    if (null != mPacketReader) {
                        mPacketReader.addRequest(request);
                    }
                }
                mPacketWriter.writePacket(packet);
            } else {
                Log.e(TAG, "Send error: PacketWriter == null");
            }
        }
    }

    public void setSocketCallback(ISocketCallback socketCallback) {
        this.mSocketCallback = socketCallback;
    }

    public interface ISocketCallback {
        int CONNECT_ERROR_CODE_CONNECT_UNKNOWN = 1;
        int CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE = 2;
        int CONNECT_ERROR_CODE_CONNECT_TIMEOUT = 3;
        int CONNECT_ERROR_CODE_HAND_SHAKE_NOT_COMPLETE = 4;
        int READ_ERROR_CODE_CONNECT_UNKNOWN = 101;
        int WRITE_ERROR_CODE_CONNECT_UNKNOWN = 102;
        int READ_ERROR_CODE_RECEIVE_LENGTH_TOO_BIG = 103;

        void onConnected();

        void onDisConnected();

        void onReceiveMsg(String msg);

        void onConnectError(int errorCode, Exception e);

        void onReadAndWriteError(int errorCode);

        void onWriterIdle();

        void onReaderIdle();

        class KageSocketCallbackThreadHandler extends Handler {
            private static HandlerThread mHandlerThread;
            private static IProtocolHandler.KageProtocolThreadHandler mHandler;

            public KageSocketCallbackThreadHandler(Looper looper) {
                super(looper);
            }

            public static IProtocolHandler.KageProtocolThreadHandler getInstance() {
                if (null == mHandler) {
                    synchronized (IProtocolHandler.KageProtocolThreadHandler.class) {
                        if (null == mHandler) {
                            if (null == mHandlerThread) {
                                mHandlerThread = new HandlerThread(KageSocketCallbackThreadHandler.class.getSimpleName());
                                mHandlerThread.start();
                            }
                            mHandler = new IProtocolHandler.KageProtocolThreadHandler(mHandlerThread.getLooper());
                        }
                    }
                }
                return mHandler;
            }
        }
    }

    private boolean isConnected() {
        return null != mSocket && null != mIn && null != mOut && null != mPacketWriter && null != mPacketReader;
    }
}

File: app/src/main/java/com/absinthe/kage/connect/tcp/KageSocket.kt
package com.absinthe.kage.connect.tcp

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import com.absinthe.kage.connect.protocol.IProtocolHandler.KageProtocolThreadHandler
import com.absinthe.kage.connect.tcp.KageSocket.ISocketCallback.KageSocketCallbackThreadHandler
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.*

class KageSocket {

    private var mSocket: Socket? = null
    private var mSocketCallback: ISocketCallback? = null
    private var mIn: DataInputStream? = null
    private var mOut: DataOutputStream? = null
    private var mPacketWriter: IPacketWriter? = null
    private var mPacketReader: IPacketReader? = null

    fun connect(ip: String?, port: Int, timeout: Int) {
        synchronized(KageSocket::class.java) {
            if (mSocket == null) {
                ConnectThread(ip, port, timeout).start()
            } else {
                Log.e(TAG, "Socket is already exist")
            }
        }
    }

    private inner class ConnectThread(private val ip: String?, private val port: Int, private val timeout: Int) : Thread() {

        override fun run() {
            synchronized(KageSocket::class.java) {
                if (mSocket == null) {
                    try {
                        mSocket = Socket()

                        val endPoint: SocketAddress = InetSocketAddress(ip, port)
                        mSocket!!.connect(endPoint, timeout)
                        mSocket!!.keepAlive = true
                        mIn = DataInputStream(mSocket!!.getInputStream())
                        mOut = DataOutputStream(mSocket!!.getOutputStream())
                        mPacketWriter = PacketWriter(mOut!!, mSocketCallback)
                        mPacketReader = PacketReader(mIn, mSocketCallback)

                        KageSocketCallbackThreadHandler.instance!!.post {
                            if (mSocketCallback != null) {
                                mSocketCallback!!.onConnected()
                            }
                        }
                    } catch (e: Exception) {
                        Log.i(TAG, "Socket connection Exception: $e")
                        mSocket = null
                        when (e) {
                            is SocketTimeoutException -> {
                                KageSocketCallbackThreadHandler.instance!!.post {
                                    if (mSocketCallback != null) {
                                        mSocketCallback!!.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_TIMEOUT, e)
                                    }
                                }
                            }
                            is ConnectException -> {
                                KageSocketCallbackThreadHandler.instance!!.post {
                                    if (mSocketCallback != null) {
                                        mSocketCallback!!.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE, e)
                                    }
                                }
                            }
                            else -> {
                                KageSocketCallbackThreadHandler.instance!!.post {
                                    if (mSocketCallback != null) {
                                        mSocketCallback!!.onConnectError(ISocketCallback.CONNECT_ERROR_CODE_CONNECT_UNKNOWN, e)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

    }

    fun disConnect(): Boolean {
        synchronized(KageSocket::class.java) {
            return if (mSocket != null && mSocket!!.isConnected) {
                try {
                    mIn!!.close()
                    mOut!!.close()
                    mSocket!!.close()
                    mIn = null
                    mOut = null
                    mSocket = null
                    if (null != mPacketWriter) {
                        mPacketWriter!!.shutdown()
                        mPacketWriter = null
                    }
                    if (null != mPacketReader) {
                        mPacketReader!!.shutdown()
                        mPacketReader = null
                    }
                    KageSocketCallbackThreadHandler.instance!!.post {
                        if (mSocketCallback != null) {
                            mSocketCallback!!.onDisConnected()
                        }
                    }
                    true
                } catch (e: IOException) {
                    Log.e(TAG, e.toString())
                    false
                }
            } else false
        }
    }

    fun send(packet: Packet?) {
        synchronized(KageSocket::class.java) {
            if (mPacketWriter != null) {
                if (packet is Request) {
                    packet.id = System.currentTimeMillis().toString()
                    mPacketReader?.addRequest(packet)
                }
                mPacketWriter?.writePacket(packet)
            } else {
                Log.e(TAG, "Send error: PacketWriter == null")
            }
        }
    }

    fun setSocketCallback(socketCallback: ISocketCallback?) {
        mSocketCallback = socketCallback
    }

    interface ISocketCallback {
        fun onConnected()
        fun onDisConnected()
        fun onReceiveMsg(msg: String)
        fun onConnectError(errorCode: Int, e: Exception)
        fun onReadAndWriteError(errorCode: Int)
        fun onWriterIdle()
        fun onReaderIdle()

        class KageSocketCallbackThreadHandler(looper: Looper) : Handler(looper) {
            companion object {
                private var mHandlerThread: HandlerThread? = null
                private var mHandler: KageProtocolThreadHandler? = null
                val instance: KageProtocolThreadHandler?
                    get() {
                        if (null == mHandler) {
                            synchronized(KageProtocolThreadHandler::class.java) {
                                if (null == mHandler) {
                                    if (null == mHandlerThread) {
                                        mHandlerThread = HandlerThread(KageSocketCallbackThreadHandler::class.java.simpleName)
                                        mHandlerThread!!.start()
                                    }
                                    mHandler = KageProtocolThreadHandler(mHandlerThread!!.looper)
                                }
                            }
                        }
                        return mHandler
                    }
            }
        }

        companion object {
            const val CONNECT_ERROR_CODE_CONNECT_UNKNOWN = 1
            const val CONNECT_ERROR_CODE_CONNECT_IP_OR_PORT_UNREACHABLE = 2
            const val CONNECT_ERROR_CODE_CONNECT_TIMEOUT = 3
            const val CONNECT_ERROR_CODE_HAND_SHAKE_NOT_COMPLETE = 4
            const val READ_ERROR_CODE_CONNECT_UNKNOWN = 101
            const val WRITE_ERROR_CODE_CONNECT_UNKNOWN = 102
            const val READ_ERROR_CODE_RECEIVE_LENGTH_TOO_BIG = 103
        }
    }

    private val isConnected: Boolean
        get() = null != mSocket
                && null != mIn
                && null != mOut
                && null != mPacketWriter
                && null != mPacketReader

    companion object {
        private val TAG = KageSocket::class.java.simpleName
    }
}